File: atp-tour/src/main/java/com/silab/atptour/service/UserService.java
package com.silab.atptour.service;

import com.silab.atptour.entity.User;
import com.silab.atptour.exceptions.AtpEntityExistsException;
import com.silab.atptour.exceptions.AtpEntityNotFoundException;
import org.springframework.security.core.userdetails.UserDetailsService;

/**
 * Represents a service containing all the logic for managing user data
 * 
 * @author Lazar
 */
public interface UserService extends UserDetailsService {

    /**
     * Gets the user with the given credentials from the database
     * 
     * @param username A string representing user's username
     * @param password A string representing user's password
     * 
     * @return A {@link User} with the given credentials
     * 
     * @throws AtpEntityNotFoundException If there's no {@link User} under the given credentials to be found
     */
    public User login(String username, String password) throws AtpEntityNotFoundException;

    /**
     * Adds a new {@link User} to the database
     * 
     * @param user A {@link User} object to be added
     * 
     * @return An added {@link User}
     * 
     * @throws AtpEntityExistsException If a {@link User} with same username already exists in the database
     */
    public User register(User user) throws AtpEntityExistsException;

    /**
     * Adds user's data in the database
     * 
     * @param user A {@link User} object to be updated
     * 
     * @return An updated {@link User}
     * 
     * @throws AtpEntityNotFoundException If the given {@link User} doesn't exist in the database
     * @throws AtpEntityExistsException If a {@link User} with same username already exists in the database
     */
    public User updateUser(User user) throws AtpEntityNotFoundException, AtpEntityExistsException;
}

File: atp-tour-angular/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { User } from '../../models/user.model';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { environment } from 'src/environments/environment';

/**
 * Service for user data management
 * 
 * @author Lazar
 */
@Injectable({
  providedIn: 'root'
})
export class UserService {

  private API_URL: string = environment.API_URL;

  /**
   * @constructor
   * 
   * @param {HttpClient} httpClient 
   */
  constructor(private httpClient: HttpClient) { }

  /**
   * Gets user from the database based on the credentials
   * 
   * @param {User} user User to be found
   * 
   * @returns {Observable<User>} Found user
   */
  public loginUser(user: User): Observable<User> {
    return this.httpClient.post<User>(this.API_URL + "/user/login", user);
  }

  /**
   * Adds new user to the database
   * 
   * @param {User} user User to be added
   * 
   * @returns {Observable<User>} Added user
   */
  public registerUser(user: User): Observable<User> {
    return this.httpClient.post<User>(this.API_URL + "/user/register", user);
  }

}


